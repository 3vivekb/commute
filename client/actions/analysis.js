import moment from 'moment'
import {push} from 'react-router-redux'
import {createAction} from 'redux-actions'
import uuid from 'uuid'

const addAnalysis = createAction('add analysis')
export const createAnalysis = (newAnalysis, organizationId) => {
  newAnalysis.id = uuid.v4()
  newAnalysis.lastRunDateTime = moment().unix()
  newAnalysis.trips = []
  return [
    addAnalysis({ analysis: newAnalysis, organizationId }),
    push(`/organizations/${organizationId}/analysis/${newAnalysis.id}`)
  ] // TODO save to server
}

const deleteLocally = createAction('delete analysis')
/* const deleteOnServer = (id) =>
  serverAction({
    url: `/api/analysis/${id}`,
    options: {
      method: 'delete'
    }
  }) */ // TODO delete on server
export const deleteAnalysis = (id, organzationId) => [
  deleteLocally({ analysisId: id, organzationId }),
  push(`/organizations/${organzationId}`)
]
