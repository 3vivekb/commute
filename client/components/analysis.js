import React, {Component} from 'react'

export default class Analysis extends Component {
  static propTypes = {}

  render () {
    return <h1>Analysis</h1>
  }
}
