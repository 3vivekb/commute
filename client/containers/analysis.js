import {connect} from 'react-redux'

import analysisActions from '../actions/analysis'
import _histogram from '../components/analysis/histogram'
import _individuals from '../components/analysis/individuals'
import _possibilities from '../components/analysis/possibilities'
import _summary from '../components/analysis/summary'

function mapStateToProps (state, props) {
  const {analysis, commuter, group, site} = state
  const currentAnalysis = analysis[props.params.analysisId]
  return {
    analysis: currentAnalysis,
    commuterStore: commuter,
    groupName: group[currentAnalysis.groupId].name,
    siteName: site[currentAnalysis.siteId].name
  }
}

function mapDispatchToProps (dispatch) {
  return {
    deleteAnalysis: (opts) => dispatch(analysisActions.delete(opts)),
    loadAnalysis: (opts) => dispatch(analysisActions.loadOne(opts))
  }
}

export const Histogram = connect(mapStateToProps)(_histogram)
export const Individuals = connect(mapStateToProps)(_individuals)
export const Possibilities = connect(mapStateToProps)(_possibilities)
export const Summary = connect(mapStateToProps, mapDispatchToProps)(_summary)
