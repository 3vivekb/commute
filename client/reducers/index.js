import {handleActions} from 'redux-actions'

import * as commuter from './commuter'
import * as multiSite from './multi-site'
import * as site from './site'
import * as user from './user'

export default {
  commuter: handleActions(commuter.reducers, commuter.initialState),
  multiSite: handleActions(multiSite.reducers, multiSite.initialState),
  site: handleActions(site.reducers, site.initialState),
  user: handleActions(user.reducers, user.initialState)
}
