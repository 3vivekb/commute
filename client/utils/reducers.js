import update from 'react-addons-update'

import {addToEntityMap, addEntitiesToEntityMap, deleteFromEntityMap, deleteManyFromEntityMap} from './entities'

function addChildEntityToParent (parentMap, newEntity, parentIdField, childPluralName) {
  return update(parentMap, {
    [newEntity[parentIdField]]: {
      [childPluralName]: {
        $push: [newEntity._id]
      }
    }
  })
}

/**
 * Make handlers for children creation and deletion events
 *
 * @param  {Object} cfg  An configuration object structured as follows:
 *   - childPluralName (String): Plural name of child (parent's child id array)
 *   - childSingularName (String): Singular name of child (for making reducer title)
 *   - parentIdField (String): Id field of parent in child entity
 * @return {Object}     Reducers
 */
export function makeChildrenHandlers (cfg) {
  const reducers = {}
  reducers[`add ${cfg.childSingularName}`] = function (state, action) {
    const newEntity = action.payload
    return addChildEntityToParent(
      state,
      newEntity,
      cfg.parentIdField,
      cfg.childPluralName
    )
  }

  reducers[`add many ${cfg.childPluralName}`] = function (state, action) {
    const newEntities = action.payload
    let updatedState = state
    newEntities.forEach(newEntity => {
      updatedState = addChildEntityToParent(
        updatedState,
        newEntity,
        cfg.parentIdField,
        cfg.childPluralName
      )
    })
    return updatedState
  }

  reducers[`delete ${cfg.childSingularName}`] = function (state, action) {
    const entity = action.payload
    const childIds = state[entity[cfg.parentIdField]][cfg.childPluralName]
    return update(state, {
      [entity[cfg.parentIdField]]: {
        [cfg.childPluralName]: {
          $set: childIds.filter((id) => id !== entity._id)
        }
      }
    })
  }

  return reducers
}

/**
 * Make generic reducers for an entity type
 *
 * @param  {Object} cfg  An object with name and handlers parameters structured as follows:
 *   - name: must have singular and plural keys with the corresponding strings
 *   - handlers: An array of strings of handlers.  Possible values: ['add', 'add many', 'delete', 'set', 'set many']
 * @return {Object}      The generated reducers
 */
export function makeGenericReducerHandlers (cfg) {
  const reducers = {
    'log out' (state, action) {
      return {}
    }
  }
  if (cfg.handlers.indexOf('add') !== -1) {
    reducers[`add ${cfg.name.singular}`] = function (state, action) {
      return addToEntityMap(state, action.payload)
    }
  }

  if (cfg.handlers.indexOf('add many') !== -1) {
    reducers[`add many ${cfg.name.plural}`] = function (state, action) {
      return addEntitiesToEntityMap(state, action.payload)
    }
  }

  if (cfg.handlers.indexOf('delete') !== -1) {
    reducers[`delete ${cfg.name.singular}`] = function (state, action) {
      return deleteFromEntityMap(state, action.payload._id)
    }
  }

  if (cfg.handlers.indexOf('delete many') !== -1) {
    reducers[`delete many ${cfg.name.plural}`] = function (state, action) {
      return deleteManyFromEntityMap(state, action.payload)
    }
  }

  if (cfg.handlers.indexOf('set') !== -1) {
    reducers[`set ${cfg.name.singular}`] = function (state, action) {
      return addToEntityMap(state, action.payload)
    }
  }

  if (cfg.handlers.indexOf('set many') !== -1) {
    reducers[`set ${cfg.name.plural}`] = function (state, action) {
      return addEntitiesToEntityMap(state, action.payload)
    }
  }

  return reducers
}
