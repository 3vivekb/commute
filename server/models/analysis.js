const {Schema} = require('mongoose')

const db = require('../db')
const Commuter = db.model('Commuter', require('./commuter'))
const Group = db.model('Group', require('./group'))
const trashPlugin = require('./plugins/trash')
const Site = db.model('Site', require('./site'))
const Trip = db.model('Trip', require('./trip'))
const dbUtils = require('../utils/db')
const later = require('../utils/later')

const schema = new Schema({
  calculationStatus: {
    default: 'calculating',
    required: true,
    type: String
  },
  groupId: {
    ref: 'Group',
    required: true,
    type: Schema.Types.ObjectId
  },
  name: {
    required: true,
    type: String
  },
  numCommuters: {
    default: 0,
    required: true,
    type: Number
  },
  organizationId: {
    ref: 'Organization',
    required: true,
    type: Schema.Types.ObjectId
  },
  siteId: {
    ref: 'Site',
    required: true,
    type: Schema.Types.ObjectId
  },
  summary: {
    // fields
    avgTravelTime: Number,
    avgDistance: Number,
    savingsPerTrip: Number,
    savingsPerTripYear: Number,
    savingsTotalPerDay: Number,
    savingsTotalPerYear: Number,

    // mongoose
    default: {
      avgTravelTime: 0,
      avgDistance: 0,
      savingsPerTrip: 0,
      savingsPerTripYear: 0,
      savingsTotalPerDay: 0,
      savingsTotalPerYear: 0
    },
    required: true,
    type: Schema.Types.Mixed
  }
})

schema.plugin(trashPlugin)

schema.pre('save', dbUtils.makeCascadeDeleteModelsFn('analysisId', [Trip]))

schema.pre('save', true, function (next, done) {
  next()

  const self = this

  if (['skipCalculation', 'calculated'].indexOf(this.calculationStatus) !== -1 ||
    this.trashed) return done()

  const profiler = require('../utils/profiler')

  // initiate profile analysis of all commuters in group
  // find all commuters via groupId
  // make sure group isn't trashed
  const ensureGroupExists = Group.count({ _id: this.groupId, trashed: undefined }).exec()
  const findCommuters = Commuter.find({ groupId: this.groupId, trashed: undefined }).exec()
  const getSite = Site.findOne({ _id: this.siteId, trashed: undefined }).exec()
  Promise.all([findCommuters, getSite, ensureGroupExists])
    .then((results) => {
      // verify that group, commuters and site exist
      const commuters = results[0]
      const site = results[1]
      const groupCount = results[2]

      if (commuters.length === 0) return done('No commuters found!')
      if (!site) return done('Site not found!')
      if (groupCount === 0) return done('Group not found!')

      self.numCommuters = commuters.length

      later(() => {
        profiler({
          analysisId: self._id,
          commuters,
          site
        })
      })
      done()
    })
    .catch((err) => {
      console.error(err)
      self.calculationStatus = 'error'
      done()
    })
})

module.exports = schema
