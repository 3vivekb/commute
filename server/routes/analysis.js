const models = require('../models')
const makeRestEndpoints = require('../utils/restEndpoints')

module.exports = function makeRoutes (app) {
  makeRestEndpoints(app,
    {
      childModels: [{
        foreignKey: 'anlaysisId',
        key: 'trips',
        model: models.Trip
      }],
      commands: {
        'Collection GET': {},
        'Collection POST': {},
        'GET': {},
        'DELETE': {}
      },
      model: models.Analysis,
      name: 'analysis'
    }
  )
}
