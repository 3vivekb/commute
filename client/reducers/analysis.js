import update from 'react-addons-update'

import analysisDefaults from '../utils/analysisDefaults'
import {fixedRound} from '../utils/common'
import {addToEntityMap} from '../utils/entities'
import {makeGenericReducerHandlers} from '../utils/reducers'

const METRICS = Object.keys(analysisDefaults.metrics)

export const reducers = makeGenericReducerHandlers({
  handlers: ['add', 'delete', 'set many'],
  name: {
    singular: 'analysis',
    plural: 'analyses'
  }
})

reducers['receive mock calculated trips'] = function (state, action) {
  const {analysisId, trips} = action.payload
  const affectedAnalysis = state[analysisId]
  let updatedAnalysis = update(affectedAnalysis, {
    trips: { $set: trips }
  })

  // calculate array of values for each metric
  const vals = {}
  const modes = Object.keys(trips[0]).filter((mode) => (['commuterId']).indexOf(mode) === -1)
  modes.forEach((mode) => {
    vals[mode] = {
      cost: [],
      distance: [],
      time: []
    }
  })
  trips.forEach((trip) =>
    modes.forEach((mode) =>
      METRICS.forEach((metric) => {
        vals[mode][metric].push(trip[mode][metric])
      })
    )
  )

  // sort arrays
  modes.forEach((mode) =>
    METRICS.forEach((metric) => {
      vals[mode][metric].sort()
    })
  )

  updatedAnalysis = update(updatedAnalysis, {
    tripVals: { $set: vals }
  })

  // calculate analysis summary metrics
  // Potential Savings is an overall average of the difference in cost between
  // the most likely trip and the driving trip
  const len = trips.length
  let totalCostDifference = 0
  let totalTravelTime = 0
  let totalDistance = 0
  trips.forEach((trip) => {
    totalCostDifference += trip.car.cost - trip.mostLikely.cost
    totalTravelTime += trip.mostLikely.time
    totalDistance += trip.mostLikely.distance
  })

  updatedAnalysis = update(updatedAnalysis, {
    summary: {
      $set: {
        avgTravelTime: fixedRound(totalTravelTime / len),
        avgDistance: fixedRound(totalDistance / len),
        savingsPerTrip: fixedRound(totalCostDifference / len),
        savingsPerTripYear: fixedRound(totalCostDifference / len * 260),
        savingsTotalPerDay: fixedRound(totalCostDifference),
        savingsTotalPerYear: fixedRound(totalCostDifference * 260)
      }
    }
  })

  return addToEntityMap(state, updatedAnalysis)
}

export const initialState = {}
