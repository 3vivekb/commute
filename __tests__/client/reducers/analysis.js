/* globals describe */

import {mockAnalysis, mockStores} from '../../test-utils/mock-data'
import {makeGenericReducerTestCases} from '../../test-utils/reducers'

import * as analysis from '../../../client/reducers/analysis'

describe('client > reducers > analysis', () => {
  makeGenericReducerTestCases({
    handlers: {
      add: {
        initialState: analysis.initialState,
        payload: mockAnalysis
      },
      delete: {
        initialState: mockStores.withAnalysisRun.analysis,
        payload: mockAnalysis
      },
      'set many': {
        initialState: analysis.initialState,
        payload: [mockAnalysis]
      }
    },
    initialState: analysis.initialState,
    name: {
      plural: 'analyses',
      singular: 'analysis'
    },
    reducers: analysis.reducers
  })
})
