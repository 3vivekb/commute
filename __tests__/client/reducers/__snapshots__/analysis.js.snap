exports[`client > reducers > analysis generic reducer tests should handle add analysis 1`] = `
Object {
  "analysis-2": Object {
    "_id": "analysis-2",
    "groupId": "group-2",
    "lastRunDateTime": 1477697490,
    "name": "An Analysis",
    "organizationId": "organization-2",
    "siteId": "site-2",
    "summary": Object {
      "avgDistance": 12.34,
      "avgTravelTime": 1234,
      "savingsPerTrip": 1.23,
      "savingsPerTripYear": 319.8,
      "savingsTotalPerDay": 1.23,
      "savingsTotalPerYear": 319.8,
    },
    "tripVals": Object {
      "bike": Object {
        "cost": Array [
          0.12,
        ],
        "distance": Array [
          23,
        ],
        "time": Array [
          2345,
        ],
      },
      "car": Object {
        "cost": Array [
          6.78,
        ],
        "distance": Array [
          18,
        ],
        "time": Array [
          1234,
        ],
      },
      "transit": Object {
        "cost": Array [
          3.45,
        ],
        "distance": Array [
          30,
        ],
        "time": Array [
          3456,
        ],
      },
      "walk": Object {
        "cost": Array [
          0,
        ],
        "distance": Array [
          19,
        ],
        "time": Array [
          6789,
        ],
      },
    },
    "trips": Array [
      Object {
        "bike": Object {
          "cost": 0.12,
          "distance": 23,
          "polygon": "encoded",
          "time": 2345,
        },
        "car": Object {
          "cost": 6.78,
          "distance": 18,
          "polygon": "encoded",
          "time": 1234,
        },
        "commuterId": "commuter-2",
        "mostLikely": Object {
          "cost": 3.45,
          "distance": 30,
          "mode": "transit",
          "polygon": "encoded",
          "time": 3456,
        },
        "transit": Object {
          "cost": 3.45,
          "distance": 30,
          "polygon": "encoded",
          "time": 3456,
        },
        "walk": Object {
          "cost": 0,
          "distance": 19,
          "polygon": "encoded",
          "time": 6789,
        },
      },
    ],
  },
}
`;

exports[`client > reducers > analysis generic reducer tests should handle delete analysis 1`] = `Object {}`;

exports[`client > reducers > analysis generic reducer tests should handle set analyses 1`] = `
Object {
  "analysis-2": Object {
    "_id": "analysis-2",
    "groupId": "group-2",
    "lastRunDateTime": 1477697490,
    "name": "An Analysis",
    "organizationId": "organization-2",
    "siteId": "site-2",
    "summary": Object {
      "avgDistance": 12.34,
      "avgTravelTime": 1234,
      "savingsPerTrip": 1.23,
      "savingsPerTripYear": 319.8,
      "savingsTotalPerDay": 1.23,
      "savingsTotalPerYear": 319.8,
    },
    "tripVals": Object {
      "bike": Object {
        "cost": Array [
          0.12,
        ],
        "distance": Array [
          23,
        ],
        "time": Array [
          2345,
        ],
      },
      "car": Object {
        "cost": Array [
          6.78,
        ],
        "distance": Array [
          18,
        ],
        "time": Array [
          1234,
        ],
      },
      "transit": Object {
        "cost": Array [
          3.45,
        ],
        "distance": Array [
          30,
        ],
        "time": Array [
          3456,
        ],
      },
      "walk": Object {
        "cost": Array [
          0,
        ],
        "distance": Array [
          19,
        ],
        "time": Array [
          6789,
        ],
      },
    },
    "trips": Array [
      Object {
        "bike": Object {
          "cost": 0.12,
          "distance": 23,
          "polygon": "encoded",
          "time": 2345,
        },
        "car": Object {
          "cost": 6.78,
          "distance": 18,
          "polygon": "encoded",
          "time": 1234,
        },
        "commuterId": "commuter-2",
        "mostLikely": Object {
          "cost": 3.45,
          "distance": 30,
          "mode": "transit",
          "polygon": "encoded",
          "time": 3456,
        },
        "transit": Object {
          "cost": 3.45,
          "distance": 30,
          "polygon": "encoded",
          "time": 3456,
        },
        "walk": Object {
          "cost": 0,
          "distance": 19,
          "polygon": "encoded",
          "time": 6789,
        },
      },
    ],
  },
}
`;
